Provide public API for PulseAudio stream renaming

tmiw opened the pull request:
Integrates the changes here into the main PR for PulseAudio support (https://github.com/PortAudio/portaudio/pull/336) to allow clients to rename PulseAudio streams after start.

daschuer commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -1311,3 +1311,34 @@ double GetStreamCpuLoad( PaStream * s )
+                                 0 );
+}
+
+PaError PaPulseAudio_RenameStream( PaStream *s, char *streamName )
```
```suggestion
PaError PaPulseAudio_RenameStream( PaStream *s, const char *streamName )
```

tmiw replied:
Done.

daschuer commented on include/pa_linux_pulseaudio.h:
```diff
@@ -0,0 +1,61 @@
+
+/** Renames opened stream. */
+
+void PaPulseAudio_RenameStream( PaStream *s, char *streamName );
```
```suggestion
void PaPulseAudio_RenameStream( PaStream *s, const char *streamName );
```

tmiw replied:
Done.

daschuer commented:
The stream name is initially set here: 

https://github.com/illuusio/portaudio/blob/49954213c11ee42cde1a89c5d263b90150ea0f74/src/hostapi/pulseaudio/pa_linux_pulseaudio.c#L1141 

Did you consider to provide a API that allows to set the stream name before OpenStream? We may also consider to add auto incrementing index to the the default name. This would help to distinguish two streams if a application does not make use of the host API specific extension.

illuusio commented:
Thank you for Pull Request. As this one renames Source and Sink all together would it be convienent to have function just have for Sink and Source separated or adding sink and source to name?

daschuer commented:
I think it is required to start the commit messages with "PulseAudio:"

illuusio commented:
And sorry to nack in 2021 but indent is not correct..

illuusio commented:
> The stream name is initially set here:
> 
> https://github.com/illuusio/portaudio/blob/49954213c11ee42cde1a89c5d263b90150ea0f74/src/hostapi/pulseaudio/pa_linux_pulseaudio.c#L1141
> 
> Did you consider to provide a API that allows to set the stream name before OpenStream? We may also consider to add auto incrementing index to the the default name. This would help to distinguish two streams if a application does not make use of the host API specific extension.

Would it be suicide at trying to get name on PaStreamParameters or use PaStreamParameters hostApiSpecificStreamInfo for giving name for stream at the start. Then we could also pass those Props on that (which seems to be 'mandatory' but no implementation is not using them). Them stream name could be changed (If wanted after that with this..)

tmiw commented:
@illuusio, the indent in the modified lines seems to match what the rest of each associated file is doing. Can you clarify as to what should be updated?

Also, it's probably a good idea to have separate functions simply so that we can actually wait for the callback to be called the same number of times as the number of streams involved. I'll work on this later today.

@daschuer, the commit messages have been updated accordingly. I actually did consider trying to get the stream names changed before open earlier but didn't get too far; it's made more complicated by none of the other hostapis using PaStreamInfo as previously mentioned. I'll research some more on this.

illuusio commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -1311,3 +1311,34 @@ double GetStreamCpuLoad( PaStream * s )
+        pa_stream_set_name(stream->outStream, streamName, RenameStreamCb, stream);
+    }
+    pa_threaded_mainloop_unlock(stream->mainloop);
+
```
Space here pa_threaded_mainloop_unlock( stream->mainloop );

illuusio commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -1311,3 +1311,34 @@ double GetStreamCpuLoad( PaStream * s )
+    }
+    else if (stream->outStream != NULL)
+    {
+        pa_stream_set_name(stream->outStream, streamName, RenameStreamCb, stream);
```
Spaces around pa_stream_set_name( stream->outStream, streamName, RenameStreamCb, stream );

illuusio commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -1311,3 +1311,34 @@ double GetStreamCpuLoad( PaStream * s )
+    if (stream->inStream != NULL)
+    {
+        pa_stream_set_name(stream->inStream, streamName, RenameStreamCb, stream);
+    }
```
spaces pa_stream_set_name( stream->inStream, streamName, RenameStreamCb, stream );

illuusio commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -1311,3 +1311,34 @@ double GetStreamCpuLoad( PaStream * s )
+    PaError result = paNoError;
+    
+    pa_threaded_mainloop_lock(stream->mainloop);
+    if (stream->inStream != NULL)
```
spaces pa_threaded_mainloop_lock( stream->mainloop );

tmiw replied:
OK, I think I got all of them. Let me know if there's any I missed.

illuusio commented:
> PaStreamInfo as previously mentioned. I'll research some more on this.

@tmiw I can add this. It's kind of for future use and now I think future has arrived :+1:

tmiw commented:
> > PaStreamInfo as previously mentioned. I'll research some more on this.
> 
> @tmiw I can add this. It's kind of for future use and now I think future has arrived 👍

Sounds good, thanks! 👍

daschuer commented:
How about tracking the steam name in this structure: https://github.com/illuusio/portaudio/blob/10c8a4050a7186bfcfcd69a3d900b9a62a946f3b/src/hostapi/pulseaudio/pa_linux_pulseaudio_internal.h#L113

And use this value here:
https://github.com/illuusio/portaudio/blob/100c25cdddfe671838f8a27768910b4122a47f1e/src/hostapi/pulseaudio/pa_linux_pulseaudio.c#L1029

I think the new funtions remain, just use the temporary storing in the case where the streams are still NULL
https://github.com/illuusio/portaudio/blob/100c25cdddfe671838f8a27768910b4122a47f1e/src/hostapi/pulseaudio/pa_linux_pulseaudio.c#L1354

I consider it much better to use instantly the final name, because Pulse is caching the old streams and I am unsure if the renaming violates the cache.

illuusio commented:
Ok I'll test this briefly. @tmiw Only thing left is Brief API docs in public header.

tmiw commented:
@illuusio, I just implemented @daschuer's suggestion and fixed the inevitable compiler errors so that the new functions can be called prior to starting the stream. This hopefully won't interfere with any PaStreamParameters work that's done.  

I also added public API documentation in the header file as appropriate. Let me know if the formatting needs to be tweaked on those.

illuusio commented:
@tmiw starting to look ready to merge. I try to find time for testing today or tomorrow so I can merge these PR's in

illuusio commented:
@tmiw are you Ok with squash merge or do you prefer to have all commits shown?

illuusio commented on include/pa_linux_pulseaudio.h:
```diff
@@ -0,0 +1,79 @@
+/*
+ * $Id$
+ * PortAudio Portable Real-Time Audio Library
+ * ALSA-specific extensions
```
Not ALSA but PortAudio

tmiw replied:
Fixed.

illuusio commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -1311,3 +1334,82 @@ double GetStreamCpuLoad( PaStream * s )
+    PaUtil_FreeMemory( stream->sinkStreamName );
+    stream->sinkStreamName = newStreamName;
+
+    if ( stream->outStream == NULL )
```
Actually we shouldn't change name if outStream is NULL. I consider it as error.

tmiw replied:
The use case is to be able to set the name after Pa_OpenStream but prior to Pa_StartStream. The stream would be NULL in that instance, correct?

illuusio replied:
Ok if this a case then 'PaPulseAudio_Initialize'-function should set stream->outStream to NULL (also inStream) as they marked NULL at  OpenStream (https://github.com/illuusio/portaudio/blob/10c8a4050a7186bfcfcd69a3d900b9a62a946f3b/src/hostapi/pulseaudio/pa_linux_pulseaudio.c#L958). Another problem is that you have to call (or if it's possible to make otherwise.. then let me know) OpenStream to initialize PaStream-struct so I don't know how we getting into this situation?

tmiw replied:
I took another look. OpenStream() initially sets both inStream and outStream to NULL. Depending on what's passed into it (inputParameters, outputParameters), one or both of those streams will be created and initialized. Thus, it's probably reasonable to return an error if the corresponding stream is NULL when renaming source or sink. I'll update.

EDIT: done.

illuusio commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -1311,3 +1334,82 @@ double GetStreamCpuLoad( PaStream * s )
+    pa_operation *op = NULL;
+    
+    /* Reallocate stream name in memory. */
+    char *newStreamName = (char*)PaUtil_AllocateMemory(strlen(streamName));
```
I think we should use strnlen like (paranoidd I know) with PAPULSEAUDIO_MAX_DEVICENAME just in case

tmiw replied:
Updated to use strnlen/strncpy with PAPULSEAUDIO_MAX_DEVICENAME.

illuusio commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -1311,3 +1334,82 @@ double GetStreamCpuLoad( PaStream * s )
+    pa_operation *op = NULL;
+
+    /* Reallocate stream name in memory. */
+    char *newStreamName = (char*)PaUtil_AllocateMemory(strlen(streamName));
```
I think we should use strnlen like (paranoid I know) with PAPULSEAUDIO_MAX_DEVICENAME just in case.

tmiw replied:
What max size would be reasonable in your opinion? 128 bytes maybe?

illuusio replied:
I think 'PAPULSEAUDIO_MAX_DEVICENAME' define should be used which now is 1024 which is overkill but I think I'll drop it to 256 bytes. Names are machine created so they can be very long.

daschuer replied:
+1 for '\0'

tmiw replied:
Updated to use sizeof() so +1 for '\0' should already be included (and to use strnlen/strncpy with PAPULSEAUDIO_MAX_DEVICENAME).

illuusio commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -1311,3 +1334,82 @@ double GetStreamCpuLoad( PaStream * s )
+    PaUtil_FreeMemory( stream->sourceStreamName );
+    stream->sourceStreamName = newStreamName;
+
+    if ( stream->inStream == NULL )
```
As in Sink this error and we should not change Source stream name if there is now stream

tmiw replied:
Updated to fail rename if stream doesn't exist.

tmiw commented:
> @tmiw are you Ok with squash merge or do you prefer to have all commits shown?

I'm fine with whatever the PortAudio project prefers.

daschuer commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -952,6 +952,19 @@ PaError OpenStream( struct PaUtilHostApiRepresentation *hostApi,
     }
 
+    /* Allocate memory for source and sink names. */
+    stream->sourceStreamName = (char*)PaUtil_AllocateMemory(strlen("Portaudio source"));
```
+1 for '\0'

tmiw replied:
Updated to use sizeof() so +1 for '\0' should already be included.

daschuer commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -952,6 +952,19 @@ PaError OpenStream( struct PaUtilHostApiRepresentation *hostApi,
 
+    /* Allocate memory for source and sink names. */
+    stream->sourceStreamName = (char*)PaUtil_AllocateMemory(strlen("Portaudio source"));
+    stream->sinkStreamName = (char*)PaUtil_AllocateMemory(strlen("Portaudio sink"));
```
The string should become a const char[] to avoid duplication. 
Than you can use sizeof().

tmiw replied:
Updated to use sizeof().

daschuer commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -1262,6 +1275,16 @@ PaError OpenStream( struct PaUtilHostApiRepresentation *hostApi,
     {
+        if ( stream->sourceStreamName )
+        {
+            PaUtil_FreeMemory( stream->sourceStreamName );
```
PaUtil_FreeMemory() has a null check internally.

tmiw replied:
Verified and removed null check since it's already being done.

illuusio commented:
Some of commit alters directly configure script which not favored thing.

tmiw commented:
@illuusio, oops, changing configure was not intended. I reverted the changes to that file.

illuusio commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -952,6 +952,22 @@ PaError OpenStream( struct PaUtilHostApiRepresentation *hostApi,
     }
 
+    /* Allocate memory for source and sink names. */
+    char defaultSourceStreamName[] = "Portaudio source";
```
I think these should be const char * as they are allocated and they are not changed anymore

tmiw replied:
Updated to use const.

illuusio commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -952,6 +952,22 @@ PaError OpenStream( struct PaUtilHostApiRepresentation *hostApi,
+    char defaultSourceStreamName[] = "Portaudio source";
+    char defaultSinkStreamName[] = "Portaudio sink";
+
+    stream->sourceStreamName = (char*)PaUtil_AllocateMemory(sizeof(defaultSourceStreamName));
```
These should be +1 for NULL at the end

tmiw replied:
sizeof() already includes the null character by virtue of the string being a constant array. Agree?

illuusio replied:
Ok I made small code like:
```
#include <stdio.h>
#include <string.h>

int main(){
   const char test1[] = "test";
   const char *test2 = "test";
   printf("Sizeof: %d strlen %d\n", sizeof(test1), strlen(test1));
   printf("Sizeof: %d strlen %d\n", sizeof(test2), strlen(test2));

return 1;
}
```
And sizeof gives correct answer and strlen gives string len so no need for +1. Btw I was wrong if used const char * then sizeof test is 8 not 5 as above gives.

tmiw replied:
That's expected as it's measuring the size of a pointer (which is 4 or 8 bytes depending on whether the system is 32 or 64 bit). Thus, you'd always want to use strlen() for those.

illuusio commented:
@daschuer If you don't find anything I think this will be merged soon

daschuer commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -1311,3 +1329,86 @@ double GetStreamCpuLoad( PaStream * s )
+    PaUtil_FreeMemory( stream->sinkStreamName );
+    stream->sinkStreamName = newStreamName;
+    
+    op = pa_stream_set_name( stream->outStream, streamName, RenameStreamCb, stream );
```
does it work if stream->outStream is still null?

tmiw replied:
Line 1385 has a NULL check for stream->outStream and returns paInvalidDevice if true.

illuusio replied:
We  have little chicken and egg problem here and below as stream is created in initialization.. so actually outStream or inStream doesn't exists before initialization. So we are in coders good will that renaming is called after that. I think we just mention this in Documentation as kind of warning and trust that people read that.

daschuer replied:
For my understanding it should work.

As soon we have a "stream" which is a Portaudio structure we can add a name to stream->sinkStreamName. 

If we have a stream->outStream, at that time, we need to update that as well by calling the pulse function. If not, it is not necessary, because the code that opens the stream will initialize it wit the stored name.

illuusio replied:
This will work only after OpenStream before that there is no allocated pointer to PaStream and you can't store anything.

daschuer replied:
Ups... 
So the whole idea is wrong to set the stream name first. We cannot do it, because we can't store it before, right?

illuusio replied:
@dashuer you are correct. This works only with streams that are already running. Thats why to have name for stream we need to to use PaStreamParameters->hostApiSpecificStreamInfo with some struct which have name and other needed stuff

tmiw replied:
For what it's worth, I haven't noticed any issues with stream renames not immediately taking effect with the current code, so it's possible that the function calls only working after Pa_OpenStream()/Pa_StartStream() is fine. Support for hostApiSpecificStreamInfo could be put into place in a future PR if desired.

Of course, it's also okay to implement it prior to merging the parent PA if you'd like :)

illuusio replied:
@tmiw Would it be easiest to get PR without this in and then start uphill battle as I know this will split minds. Personally I see this LGTM state. Or just get this in and fight the battle..

tmiw replied:
@illuusio, the current implementation seems close to ready so it might be easiest to proceed with this. However, if the other PortAudio maintainers prefer hostApiSpecificStreamInfo (in addition to or instead of this), that's a different story. If that's the case, I'm fine with deferring this and getting basic PortAudio support in.

daschuer commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -1311,3 +1329,86 @@ double GetStreamCpuLoad( PaStream * s )
+    }
+
+    /* Reallocate stream name in memory. */
+    pa_threaded_mainloop_lock( stream->mainloop );
```
we need to check if we actually have a mainloop. I think this lock should only guard the pulse calls.

tmiw replied:
The main loop is created during PortAudio initialization, correct? (PaPulseAudio_Initialize/PaPulseAudio_New)

I also remember having to move around the lock/unlock calls a bit because I was running into crash issues related to allocating/deallocating the stream names. Should we possibly be using pa_xalloc() and similar instead for those?

illuusio replied:
Yes.. as mainloop runs and if you don't lock it you will suffer.. and it there is no mainloop nothing works so.. for stream naming before init we have to come to another solution.

tmiw replied:
That seems like a different issue from what @daschuer was mentioning. For this, there shouldn't be an issue since streams can't be created/started without PortAudio having been initialized via Pa_Initialize().

daschuer commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -1311,3 +1329,86 @@ double GetStreamCpuLoad( PaStream * s )
+        return paInsufficientMemory;
+    }
+    memset(newStreamName, 0, strnlen(streamName, PAPULSEAUDIO_MAX_DEVICENAME) + 1);
+    strncpy(newStreamName, streamName, strnlen(streamName, PAPULSEAUDIO_MAX_DEVICENAME));
```
a save and fast way us using snprintf("%s" ...) it does not look like this, but the compiler is smart enough to get rid of the format parsing at runtime.

tmiw replied:
Updated.

daschuer commented on src/hostapi/pulseaudio/pa_linux_pulseaudio.c:
```diff
@@ -1026,7 +1042,7 @@ PaError OpenStream( struct PaUtilHostApiRepresentation *hostApi,
         stream->inStream =
             pa_stream_new( l_ptrPulseAudioHostApi->context,
-                           "Portaudio source",
+                           stream->sourceStreamName,
```
We can get around the memory allocation in  this function by 
```suggestion
                           stream->sourceStreamName ? stream->sourceStreamName : defaultSourceStreamName,
```

tmiw replied:
How would switching from the default stream name to the user-selected one be handled with the suggested change? Trying to deallocate static data won't go over too well.

illuusio replied:
I'll do it after this is merged as I need to add stuff to created pa_linux_pulseaudio.h or add struct that contains name and stuff for this https://www.freedesktop.org/wiki/Software/PulseAudio/Documentation/Developer/Clients/ApplicationProperties/. This struct will be passed to openStream as PaStreamParameters->hostApiSpecificStreamInfo. So if it's not NULL it will treated that there is name.

After that there will be no more new stuff that this will get in. I have been trying get this merged more than 6 years now... so I like to get rid of it someday.

tmiw replied:
I'm fine leaving this as-is pending your changes (or potentially a future PR if desired).
